sdk/typescript/src/bcs/index.js

```diff
@@ -0,0 +1,663 @@
+        }
+        /**
+         * Write a U128 value into a buffer and shift cursor position by 16.
+         * 
+         * @unimplemented
+         * @param {Number} value Value to write.
+         * @returns {this}
+         */
+        write128(value) {
+            throw 'Writing u128 values is not yet implemented';
+        }
+        /**
+         * Write a ULEB value into a buffer and shift cursor position by number of bytes
+         * written.
+         * @param {Number} value Value to write.
+         * @returns {this}
+         */
+        writeULEB(value) {
+            ulebEncode(value).forEach((el) => this.write8(el));
+            return this;
+        }
+        /**
+         * Write a vector into a buffer by first writing the vector length and then calling
+         * a callback on each passed value.
+         *
+         * @param {Array<Any>} vector Array of elements to write.
+         * @param {WriteVecCb} cb Callback to call on each element of the vector.
+         * @returns {this}
+         */
+        writeVec(vector, cb) {
+            this.writeULEB(vector.length);
+            [...vector].forEach((el, i) => cb(this, el, i, vector.length));
+            return this;
+        }
+        /**
+         * Get underlying buffer taking only value bytes (in case initial buffer size was bigger).
+         * @returns {Uint8Array} Resulting BCS.
+         */
+        toBytes() {
+            return new Uint8Array(this.dataView.buffer.slice(0, this.bytePosition));
+        }
+    }
+
+    /**
+     * The ReadVector callback.
+     * @callback WriteVecCb
+     * @param {BcsWriter} writer A writer instance to use.
+     * @param {Any} el An element to write.
+     * @param {Number} i An index of the element.
+     * @param {Number} length Total length of the vector being read.
+     */
+    
+    // Helper utility: write number as an ULEB array.
+    // Original code is taken from: https://www.npmjs.com/package/uleb128 (no longer exists)
+    function ulebEncode(num) {
+        let arr = [];
+        let len = 0;
+
+        if (num === 0) {
+            return [0];
+        }
+
+        while (num > 0) {
+            arr[len] = num & 0x7F;
+            if (num >>= 7) {
+                arr[len] |= 0x80
+            };
+            len += 1;
+        }
+
+        return arr;
+    }
+
+    // Helper utility: decode ULEB as an array of numbers.
+    // Original code is taken from: https://www.npmjs.com/package/uleb128 (no longer exists)
+    function ulebDecode(arr) {
+        let total = 0;
+        let shift = 0;
+        let len = 0;
+
+        while (true) {
+            let byte = arr[len];
+            len += 1;
+            total |= ((byte & 0x7F) << shift);
+            if ((byte & 0x80) === 0) {
+                break;
+            }
+            shift += 7;
+        }
+
+        return {
+            value: total,
+            length: len
+        };
+    }
+
+    // define constants 
+    MoveBCS.types = new Map();
```

Comment:
[Map's `iterable` constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/Map#creating_a_new_map) could make this a single statement, which may be stylistically preferable
```js
MoveBCS.types = new Map([
    ['U8', 'u8'],
    ['U32', 'u32'],
    ['U64', 'u64'],
    ['U128', 'u128'],
    ['BOOL', 'bool'],
    ['VECTOR', 'vector'],
    ['ADDRESS', 'address'],
    ['STRING', 'string']
]);
```

Reply:
It could, but `types` actually holds different mapping: `type => typeInterface { encode(), decode() }`. These `U8`, `U32` etc exist as constants if people prefer to use predefined type names constants instead of string literals like `u8` or `bool`.